utils/src/logging.rs

```diff
@@ -21,3 +29,39 @@ pub fn setup_logging() -> Result<()> {
 
     Ok(())
 }
+
+pub fn format_opt<T: Display>(opt: Option<&T>) -> String {
+    opt.map(ToString::to_string).unwrap_or("None".to_string())
+}
+
+pub fn format_hashes(hashes: &[RpoDigest]) -> String {
+    format!("[{}]", hashes.iter().map(RpoDigest::to_hex).join(", "))
+}
+
+pub fn format_input_notes(notes: &InputNotes<Nullifier>) -> String {
+    format!(
+        "{{ notes: [{}], commitment: {} }}",
+        notes.iter().map(Nullifier::to_hex).join(", "),
+        notes.commitment()
+    )
+}
+
+pub fn format_output_notes(notes: &OutputNotes<NoteEnvelope>) -> String {
+    format!(
+        "{{ notes: [{}], commitment: {} }}",
+        notes
+            .iter()
+            .map(|envelope| {
+                let metadata = envelope.metadata();
+                format!(
+                    "{{ note_id: {}, note_metadata: {{sender: {}, tag: {}, num_assets: {} }}}}",
+                    envelope.note_id().inner(),
+                    metadata.sender(),
+                    metadata.tag(),
+                    metadata.num_assets()
+                )
+            })
+            .join(", "),
+        notes.commitment()
+    )
```

Comment:
A couple of small comments here as well:

- If there are no notes, we can return "None".
- We don't need to log `commitment` field.
- `num_assets` is no longer a part of note metadata (we need to rebase this PR from `main`).